video/enhanced_samples_generator.h

```diff
@@ -22,23 +22,36 @@ class EnhancedSamplesGenerator : public WaveformGenerator {
 	
 	private:
 		std::weak_ptr<audio_sample> _sample;
+
+        int previousSample = 0;
+        int currentSample = 0;
+        double samplesPerGet = 1.0;
+        double fractionalSampleOffset = 0.0;
+
+        const int baseFrequency = 16000;
 };
 
 EnhancedSamplesGenerator::EnhancedSamplesGenerator(std::shared_ptr<audio_sample> sample)
 	: _sample(sample)
 {}
 
 void EnhancedSamplesGenerator::setFrequency(int value) {
-	// usually this will do nothing...
-	// but we'll hijack this method to allow us to reset the sample index
+	// We'll hijack this method to allow us to reset the sample index
 	// ideally we'd override the enable method, but C++ doesn't let us do that
 	if (value < 0) {
 		// rewind our sample if it's still valid
 		if (!_sample.expired()) {
 			auto samplePtr = _sample.lock();
 			samplePtr->rewind();
+
+            fractionalSampleOffset = 0.0;
+            previousSample = samplePtr->getSample();
+            currentSample = samplePtr->getSample();
```

Comment:
this confused me for a while.  this _kinda_ feels wrong, to grab the first two samples, and thus leave the sample object in a state where the next call to `getSample` will return the third sample value it contains.  that kinda feels wrong, and on thinking about this code my reaction was that this should rewind the sample again after these reads.  I suspect my confusion here is essentially the same that @HeathenUK experienced :grin:

the key of course is the `fractionalSampleOffset`, which indicates how far along the journey one is from the `previousSample` to the `currentSample`.  it's only when that ticks past 1 that that journey is completed and the next sample value read.

_arguably_ naming these two variables as `currentSample` and `nextSample` respectively _might_ have made this slightly clearer, but to be honest I'm not sure about that :grin: